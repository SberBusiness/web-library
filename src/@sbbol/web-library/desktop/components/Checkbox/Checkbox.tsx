import React, {useState, useRef, useEffect} from 'react';
import {ICheckboxProps} from '@sbbol/web-library/desktop/components/Checkbox/types';
import {classnames} from '@sbbol/web-library/common/utils/classnames/classnames';
import {EFocusSource} from '@sbbol/web-library/common/enums/EFocusSource';
import {CheckboxbulkStsIcon16} from '@sberbusiness/icons/CheckboxbulkStsIcon16';
import {CheckboxtickStsIcon16} from '@sberbusiness/icons/CheckboxtickStsIcon16';
import {isKey} from '@sbbol/web-library/desktop/utils/keyboard';

/** Компонент "Чекбокс". Используется как одиночный чекбокс с описанием. */
export const Checkbox = React.forwardRef<HTMLInputElement, ICheckboxProps>((props, ref) => {
    const {children, className, onFocus, bulk, labelAttributes, ...inputAttributes} = props;
    const [focusSource, setFocusSource] = useState(EFocusSource.NONE);
    const labelRef = useRef<HTMLLabelElement | null>(null);
    const classNames = classnames('cssClass[checkbox]', {'cssClass[focusVisible]': focusSource === EFocusSource.KEYBOARD}, className);
    const labelClassNames = classnames('cssClass[label]', {'cssClass[nonempty]': !!children}, labelAttributes?.className);

    useEffect(() => {
        if (focusSource !== EFocusSource.NONE) {
            document.addEventListener('mousedown', listenMouseDown);
            document.addEventListener('keydown', listenKeyDown);
            return () => {
                document.removeEventListener('mousedown', listenMouseDown);
                document.removeEventListener('keydown', listenKeyDown);
            };
        }
    }, [focusSource]);

    const listenMouseDown = (event: MouseEvent) => {
        if (!labelRef.current?.contains(event.target as Node)) {
            setFocusSource(EFocusSource.NONE);
        }
    };

    const listenKeyDown = (event: KeyboardEvent) => {
        const key = event.code || event.keyCode;

        if (isKey(key, 'TAB')) {
            setFocusSource(EFocusSource.NONE);
        }
    };

    const handleMouseDown = (event: React.MouseEvent<HTMLLabelElement>) => {
        if (focusSource === EFocusSource.NONE && labelRef.current !== document.activeElement) {
            setFocusSource(EFocusSource.MOUSE);
        }
        labelAttributes?.onMouseDown?.(event);
    };

    const handleFocus = (event: React.FocusEvent<HTMLInputElement>) => {
        if (focusSource === EFocusSource.NONE) {
            setFocusSource(EFocusSource.KEYBOARD);
        }
        onFocus?.(event);
    };

    const renderCheckmarkIcon = () => {
        const className = 'cssClass[checkmarkIcon]';

        return bulk ? <CheckboxbulkStsIcon16 className={className} /> : <CheckboxtickStsIcon16 className={className} />;
    };

    return (
        <label {...labelAttributes} className={labelClassNames} onMouseDown={handleMouseDown} ref={labelRef}>
            <input type="checkbox" className={classNames} onFocus={handleFocus} {...inputAttributes} ref={ref} />
            <span className="cssClass[checkboxIcon]" />
            {renderCheckmarkIcon()}
            {children && <span className="cssClass[content]">{children}</span>}
        </label>
    );
});

Checkbox.displayName = 'Checkbox';
