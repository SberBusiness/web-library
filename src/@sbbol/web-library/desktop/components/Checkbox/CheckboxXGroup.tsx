import * as React from 'react';
import {TIndentSize} from '@sbbol/web-library/desktop/common/consts/IndentConst';
import {cssClass} from '@sbbol/web-library/desktop/utils/cssClass';
import {classnames} from '@sbbol/web-library/common/utils/classnames/classnames';

/** Свойства компонента CheckboxXGroup. */
export interface ICheckboxXGroupProps extends React.HTMLAttributes<HTMLDivElement> {
    /** Размер отступа. */
    indent?: TIndentSize;
}

/** Компонент CheckboxXGroup. Группа чекбоксов с направлением по оси X. */
export const CheckboxXGroup: React.FC<ICheckboxXGroupProps> = (props) => {
    const {children, className, indent = 12, ...rest} = props;
    const classNames = classnames('cssClass[checkboxXGroup]', cssClass(`indent-${indent}`), className);

    return (
        <div className={classNames} role="group" {...rest}>
            {children}
        </div>
    );
};

CheckboxXGroup.displayName = 'CheckboxXGroup';
