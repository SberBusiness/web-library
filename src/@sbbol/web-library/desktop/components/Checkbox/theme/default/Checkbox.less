@import '../../../../common/theme/default';

.label {
    display: inline-block;
    position: relative;
    min-width: @ufs-theme-size-icon-16;
    min-height: @ufs-theme-size-icon-16;
    line-height: normal;
    cursor: pointer;
    user-select: none;

    &[hidden] {
        display: none; // Screenshot tests
    }

    &.nonempty {
        min-height: 20px;
    }
}

.checkbox {
    margin: 0;
    cursor: pointer;
    opacity: 0;
}

.checkboxIcon {
    box-sizing: border-box;
    background-color: @color-snowy;
    border: 1px solid @color-asphalt-03;
    border-radius: @ufs-theme-base;
}

.checkmarkIcon {
    transform: scale(0);
}

.content {
    display: inline-block;
    padding-left: @metric-24;
    box-sizing: border-box;
    .b1-text-main(@color-graphite-01);
}

.checkboxIcon,
.checkmarkIcon {
    pointer-events: none;
}

.checkbox,
.checkboxIcon,
.checkmarkIcon {
    position: absolute;
    width: @ufs-theme-size-icon-16;
    height: @ufs-theme-size-icon-16;
    left: 0;

    .nonempty & {
        top: 2px;
    }
}

.checkbox:checked {
    + .checkboxIcon {
        background-color: @color-malachite-02;
        border-color: @color-malachite-02;
    }

    ~ .checkmarkIcon {
        transform: scale(1);
        transition: transform 200ms ease-out;
    }
}

.checkbox:hover + .checkboxIcon {
    border-color: @color-malachite-03;
}

.checkbox.focusVisible + .checkboxIcon {
    border-color: @color-sunny-01;
}

.checkbox:disabled {
    + .checkboxIcon {
        background-color: @color-asphalt-05;
        border-color: @color-asphalt-03;
    }

    ~ .checkmarkIcon path {
        fill: @color-asphalt-03;
    }
}

.checkbox:checked:enabled:hover + .checkboxIcon {
    background-color: @color-malachite-03;
}

@media screen and (-ms-high-contrast: active), (-ms-high-contrast: none) {
    .label:hover .checkbox:enabled {
        & + .checkboxIcon {
            border-color: @color-malachite-03;
        }

        &:checked + .checkboxIcon {
            background-color: @color-malachite-03;
        }
    }
}
