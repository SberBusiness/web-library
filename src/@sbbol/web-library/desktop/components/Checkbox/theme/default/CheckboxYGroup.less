@import '../../../../common/theme/default';

.checkboxYGroup {
    display: inline-flex;
    flex-flow: column wrap;

    .label {
        &:not(:last-of-type) {
            margin-bottom: @metric-16;
        }
    }
}
