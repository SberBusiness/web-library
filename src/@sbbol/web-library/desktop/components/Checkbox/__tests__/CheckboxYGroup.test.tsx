import React from 'react';
import renderer from 'react-test-renderer';
import {allure} from '@jest/unit/allure-report';
import {CheckboxYGroup} from '@sbbol/web-library/desktop/components/Checkbox/CheckboxYGroup';
import {Checkbox} from '@sbbol/web-library/desktop/components/Checkbox/Checkbox';

jest.mock('@sberbusiness/icons/CheckboxtickStsIcon16', () => ({
    CheckboxtickStsIcon16: 'svg',
}));

jest.mock('@sberbusiness/icons/CheckboxbulkStsIcon16', () => ({
    CheckboxbulkStsIcon16: 'svg',
}));

describe('CheckboxYGroup', () => {
    beforeEach(() => {
        allure.feature('CheckboxYGroup');
    });

    it('renders correctly', () => {
        const tree = renderer
            .create(
                <CheckboxYGroup>
                    <Checkbox>Checkbox</Checkbox>
                    <Checkbox>Checkbox</Checkbox>
                    <Checkbox>Checkbox</Checkbox>
                </CheckboxYGroup>
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
});
