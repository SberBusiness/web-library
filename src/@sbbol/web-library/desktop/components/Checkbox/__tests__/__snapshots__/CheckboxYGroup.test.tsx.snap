// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckboxYGroup renders correctly 1`] = `
<div
  className="cssClass[checkboxYGroup]"
  role="group"
>
  <label
    className="cssClass[label] cssClass[nonempty]"
    onMouseDown={[Function]}
  >
    <input
      className="cssClass[checkbox]"
      onFocus={[Function]}
      type="checkbox"
    />
    <span
      className="cssClass[checkboxIcon]"
    />
    <svg
      className="cssClass[checkmarkIcon]"
    />
    <span
      className="cssClass[content]"
    >
      Checkbox
    </span>
  </label>
  <label
    className="cssClass[label] cssClass[nonempty]"
    onMouseDown={[Function]}
  >
    <input
      className="cssClass[checkbox]"
      onFocus={[Function]}
      type="checkbox"
    />
    <span
      className="cssClass[checkboxIcon]"
    />
    <svg
      className="cssClass[checkmarkIcon]"
    />
    <span
      className="cssClass[content]"
    >
      Checkbox
    </span>
  </label>
  <label
    className="cssClass[label] cssClass[nonempty]"
    onMouseDown={[Function]}
  >
    <input
      className="cssClass[checkbox]"
      onFocus={[Function]}
      type="checkbox"
    />
    <span
      className="cssClass[checkboxIcon]"
    />
    <svg
      className="cssClass[checkmarkIcon]"
    />
    <span
      className="cssClass[content]"
    >
      Checkbox
    </span>
  </label>
</div>
`;
